### ISSUE TITLE
HttpClient times out for some users while trying to get etags

### ISSUE THREAD

**Issue by dandoesgithub** [KSP-CKAN/CKAN#3950]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Operating System

Windows 10

### CKAN Version

1.34.0

### Games

KSP 1

### Game Version

KSP 1.12.5

### Did you make any manual changes to your game folder (i.e., not via CKAN)?

Yes, I deleted the mods inside of GameData

### Describe the bug

CKAN gets stuck during "Refreshing game version data", then after a while errors. This only happens with one instance.
The solution for this bug for now is searching a mod in the search bar, then refreshing

I expected it to load normally
![image](https://github.com/KSP-CKAN/CKAN/assets/54538634/db4891f9-b863-488b-aa1a-1a86f5462067)

### Steps to reproduce

Step 1:  Load an instance
Step 2: Refresh
Step 3: Get lucky (or unlucky) and get the error

**Comment by HebaruSan** [comment]
Hi @dandoesgithub, thanks for the report.

> This only happens with one instance.

To fix this we will need to know what is special about that instance. Have you perhaps added some custom repositories to it?

**Comment by dandoesgithub** [comment]
No, I haven't, it's the default repository

**Comment by HebaruSan** [comment]
OK, is there anything else about that instance that you can think of that could explain why this doesn't happen for me or anyone else up to this point?

**Comment by dandoesgithub** [comment]
Well, I deleted all the mods in the GameData folder (CKAN doesn't know that), but I don't think that's what's causing the problem. I'm not sure why it could be happening, as it was working in 1.33.2

**Comment by HebaruSan** [comment]
Which game is the special instance, KSP1 or KSP2?

**Comment by dandoesgithub** [comment]
I could take a look at the CKAN files in the particular instance

**Comment by JonnyOThan** [comment]
@dandoesgithub It might help to zip up your CKAN folder from that instance and post it here.

**Comment by dandoesgithub** [comment]
slight problem: i have deleted that instance. craaaap. i'll try to recover it with a program but i don't think i'll succeed

**Comment by HebaruSan** [comment]
Oh, in that case we won't be able to investigate this. Let us know if you do recover the instance and we can re-open it.

**Comment by dandoesgithub** [comment]
wait: the new instance that i just made has the same problem. i can send that one instead.

**Comment by HebaruSan** [comment]
It may be something with your PC rather than these instances.

**Comment by dandoesgithub** [comment]
no, every other instance works except for this one. either it actually is my PC, or it's the instance

anyway, here's the CKAN file
[CKAN.zip](https://github.com/KSP-CKAN/CKAN/files/13696535/CKAN.zip)

**Comment by HebaruSan** [comment]
Thanks, that ZIP looks fine to me. Maybe @JonnyOThan had something particular in mind.

If it's having the error during "Refreshing game version data", that gets stored in another place. Can you send us `%LOCALAPPDATA%\CKAN` as well? You can exclude the `downloads` subfolder if it's huge.

Are you using the same `ckan.exe` file for all your instances, or are you copying it into each instance? Usually that doesn't matter, but I'm trying to think of other things that might differentiate instances.

**Comment by dandoesgithub** [comment]
[CKAN.zip](https://github.com/KSP-CKAN/CKAN/files/13696672/CKAN.zip), i am using the same ckan.exe for all my instances.

**Comment by dandoesgithub** [comment]
should i try deleting the ksp-default repo then adding it again?

**Comment by HebaruSan** [comment]
Thanks, that looks all-in-order as well. I would not expect deleting and re-adding the repo to help, but I wouldn't discourage you from trying it if you think otherwise.

The next thing to try is probably capturing some logs so we can get more information about what CKAN is doing when this problem happens. If you unzip this `log4net.xml` file next to your `ckan.exe` file and then run CKAN, it should load the settings automatically and generate a `CKAN.log` file that you can share here.

[log4net.xml.zip](https://github.com/KSP-CKAN/CKAN/files/13696874/log4net.xml.zip)

```xml
<?xml version="1.0" encoding="utf-8" ?>
<log4net>
  <appender name="CkanFile" type="log4net.Appender.RollingFileAppender">
    <filter type="log4net.Filter.StringMatchFilter">
        <stringToMatch value="Finding latest available" />
        <acceptOnMatch value="false" />
    </filter>
    <filter type="log4net.Filter.StringMatchFilter">
        <stringToMatch value="Installable: " />
        <acceptOnMatch value="false" />
    </filter>
    <filter type="log4net.Filter.StringMatchFilter">
        <stringToMatch value="Already enlisted" />
        <acceptOnMatch value="false" />
    </filter>
    <filter type="log4net.Filter.StringMatchFilter">
        <stringToMatch value="Available: " />
        <acceptOnMatch value="false" />
    </filter>
    <filter type="log4net.Filter.StringMatchFilter">
        <stringToMatch value="Adding new available module " />
        <acceptOnMatch value="false" />
    </filter>
    <filter type="log4net.Filter.StringMatchFilter">
        <stringToMatch value="Adding available module " />
        <acceptOnMatch value="false" />
    </filter>
    <filter type="log4net.Filter.StringMatchFilter">
        <stringToMatch value="Adding to available module: " />
        <acceptOnMatch value="false" />
    </filter>
    <filter type="log4net.Filter.StringMatchFilter">
        <stringToMatch value="Copying " />
        <acceptOnMatch value="false" />
    </filter>
    <filter type="log4net.Filter.StringMatchFilter">
        <stringToMatch value="Writing file " />
        <acceptOnMatch value="false" />
    </filter>
    <filter type="log4net.Filter.StringMatchFilter">
        <stringToMatch value="Making directory " />
        <acceptOnMatch value="false" />
    </filter>
    <filter type="log4net.Filter.StringMatchFilter">
        <stringToMatch value="Making parent directory " />
        <acceptOnMatch value="false" />
    </filter>
    <filter type="log4net.Filter.StringMatchFilter">
        <stringToMatch value="Not registering " />
        <acceptOnMatch value="false" />
    </filter>
    <filter type="log4net.Filter.StringMatchFilter">
        <stringToMatch value="Got owners: " />
        <acceptOnMatch value="false" />
    </filter>
    <filter type="log4net.Filter.StringMatchFilter">
        <stringToMatch value="Module parsed: " />
        <acceptOnMatch value="false" />
    </filter>
    <filter type="log4net.Filter.StringMatchFilter">
        <stringToMatch value="Reading CKAN data from " />
        <acceptOnMatch value="false" />
    </filter>
    <filter type="log4net.Filter.StringMatchFilter">
        <stringToMatch value="Skipping archive entry " />
        <acceptOnMatch value="false" />
    </filter>
    <filter type="log4net.Filter.StringMatchFilter">
        <regexToMatch value="Checking compat of" />
        <acceptOnMatch value="false" />
    </filter>
    <filter type="log4net.Filter.StringMatchFilter">
        <stringToMatch value="Got contents" />
        <acceptOnMatch value="false" />
    </filter>
    <filter type="log4net.Filter.StringMatchFilter">
        <stringToMatch value="Not removing directory " />
        <acceptOnMatch value="false" />
    </filter>
    <filter type="log4net.Filter.StringMatchFilter">
        <stringToMatch value="Removing " />
        <acceptOnMatch value="false" />
    </filter>

    <file value="./CKAN.log" />
    <appendToFile value="false" />
    <rollingStyle value="Once" />
    <maxSizeRollBackups value="10" />
    <lockingModel type="log4net.Appender.FileAppender+MinimalLock" />

    <layout type="log4net.Layout.PatternLayout">
      <conversionPattern value="%6timestamp %-5level %logger  %message%newline" />
    </layout>
  </appender>
  <!-- Log everything by default -->
  <root>
    <level value="DEBUG" />
    <appender-ref ref="CkanFile" />
  </root>
  <!-- Turn down some of the noisier loggers -->
  <logger name="CKAN.CompatibilitySorter">
      <level value="INFO" />
  </logger>
  <logger name="CKAN.NetFileCache">
      <level value="INFO" />
  </logger>
  <logger name="CKAN.CkanModule">
      <level value="INFO" />
  </logger>
</log4net>
```

**Comment by Falki-git** [comment]
Just confirming that I'm getting this problem as well. But sometimes, now always. I'll try with the logging @HebaruSan suggested.

![image](https://github.com/KSP-CKAN/CKAN/assets/72734856/268ecade-44bb-44f6-bef1-a69ccdc4bd65)

**Comment by Falki-git** [comment]
Here is the log after one of this hangs.
[CKAN.log](https://github.com/KSP-CKAN/CKAN/files/13697088/CKAN.log)

**Comment by HebaruSan** [comment]
Thanks, @Falki-git. That's very strange; that log file shows CKAN completing and exiting normally. If it's the file that you  meant to send, then that may mean we don't have logging set up for wherever this problem is happening.

"One or more errors occurred" isn't even our text; it's from`AggregateException`. So some problem is happening and then the framework is obscuring it with that class. The documentation says it's used extensively by PLINQ, so this is probably related to the recent changes to parallelize various things for performance. (Or the root cause could be unrelated, but now happening in parallel and therefore obscured by PLINQ.)

Here's a test build with a first attempt to get the original problem from the `AggregateException` based on its documentation; please try it if you have time and are willing to trust that I'm not sending you malware:

- [ckan.zip](https://github.com/KSP-CKAN/CKAN/files/13697136/ckan.zip)

**Comment by Falki-git** [comment]
@HebaruSan, I tried it and I get the same behavior. I can refresh several times without a problem and then it hangs on the next refresh. Hang lasts for about a minutes and then the "task canceled" is thrown.

Here are the logs from this attempt.

[CKAN_logs.zip](https://github.com/KSP-CKAN/CKAN/files/13697210/CKAN_logs.zip)

And just a quick video of what's happening.

https://github.com/KSP-CKAN/CKAN/assets/72734856/1efcbdb2-a378-4674-b56d-a3c346d01f59

**Comment by HebaruSan** [comment]
Yeah, you're supposed to get the same behavior, all I added was handling of the `AggregateException`. That's why it now says "A task was canceled" instead of "One or more errors occurred".

The log has basically the same info. I.e., almost none:

```
107407 ERROR CKAN.GUI.Main  A task was canceled.
System.Threading.Tasks.TaskCanceledException: A task was canceled.
```

So, that's... great.

**Comment by dandoesgithub** [comment]
Deleting and re-adding the repo worked! apparently the repo was the problem!

**Comment by HebaruSan** [comment]
Well, let's see if that turns out to be the case for all users reporting this issue. I'm skeptical for various reasons.

**Comment by HebaruSan** [comment]
One more attempt to get useful info out of this exception:

- [ckan.zip](https://github.com/KSP-CKAN/CKAN/files/13697314/ckan.zip)

@Falki-git or @dandoesgithub or anyone else who can reproduce this, please try that build and let us know what it says.

**Comment by dandoesgithub** [comment]
oh, the issue is still happening on that build, deleting the default repo and re-adding it worked again

**Comment by HebaruSan** [comment]
Deleting the repo and re-adding it doesn't actually change anything, though. I think this is a race condition that has turned out differently after you've done that by coincidence.

Some support pages are saying `HttpClient` can throw this as a timeout, which on the surface makes sense since we do now use that to get the etags, but that should take over a minute and a half to time out, and that video is less than 15 seconds:

![image](https://github.com/KSP-CKAN/CKAN/assets/1559108/293ef2ee-5668-4777-9e88-7931f99a065e)

**Comment by HebaruSan** [comment]
Also @dandoesgithub the point of that build was to gather info. Can you please share a screenshot or the logs?

**Comment by dandoesgithub** [comment]
where are the logs located?

**Comment by HebaruSan** [comment]
I should say explicitly: The test builds that I have been sharing are **not attempting to fix the problem**, because we do not know what the problem is. They are solely trying to get enough information to identify what the problem is.

@dandoesgithub see above re: logs. https://github.com/KSP-CKAN/CKAN/issues/3950#issuecomment-1859219946

**Comment by dandoesgithub** [comment]
[CKAN.txt](https://github.com/KSP-CKAN/CKAN/files/13697392/CKAN.txt)

**Comment by HebaruSan** [comment]
Thanks; that looks just like Falki's log, so either you're still using the previous build, or `TaskCanceledException` doesn't provide any `InnerException`s.

```
147790 ERROR CKAN.GUI.Main  A task was canceled.
System.Threading.Tasks.TaskCanceledException: A task was canceled.
```

Side note, the 100-second default timeout seems to be confirmed by inspecting an `HttpClient` object:

```
csharp> LoadAssembly("System.Net.Http");
csharp> var hc = new System.Net.Http.HttpClient();
csharp> hc.Timeout;
00:01:40
```

**Comment by Falki-git** [comment]
@HebaruSan, here is the log for the newest ckan.exe you sent:
[CKAN_log2.zip](https://github.com/KSP-CKAN/CKAN/files/13697512/CKAN_log2.zip)

Regarding the video, I should have been explicit also, I cut it down to 15 seconds so you don't need to look at a static screen, but from clicking Refresh to the timeout it takes a minute or so (likely 100 secs as you said).

**Comment by HebaruSan** [comment]
@dandoesgithub and @Falki-git, here's a new test build for you to try:

[ckan.zip](https://github.com/KSP-CKAN/CKAN/files/13697516/ckan.zip)

This one does new things: first it prints the value of `HttpClient.Timeout`, and then it catches, logs, and ignores any exceptions it throws while trying to get the etags. This _**might**_ help with the problem you're having, but whether it does or not, it's important that we get the log file so we can determine the right long term fix for the real builds.

> Regarding the video, I should have been explicit also, I cut it down to 15 seconds so you don't need to look at a static screen, but from clicking Refresh to the timeout it takes a minute or so (likely 100 secs as you said).

Thanks, that makes a lot more sense! That makes me a lot more confident concluding that this is in fact `HttpClient` timing out. I built and typed the above before you posted this, but I think it's still worth capturing those logs.

**Comment by Falki-git** [comment]
And here's the log. It's more verbose this time.

[CKAN_logs3.zip](https://github.com/KSP-CKAN/CKAN/files/13697539/CKAN_logs3.zip)

**Comment by HebaruSan** [comment]
> And here's the log.

Thanks! Yup, `HttpClient.SendAsync` is definitely throwing.

```
 31572 INFO  CKAN.Net  HttpClient.Timeout is 00:01:40
131575 WARN  CKAN.Net  Failed to get etag for https://github.com/KSP-CKAN/KSP2-CKAN-meta/archive/main.tar.gz
System.AggregateException: One or more errors occurred. ---> System.Threading.Tasks.TaskCanceledException: A task was canceled.
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task`1.GetResultCore(Boolean waitCompletionNotification)
   at CKAN.Net.CurrentETag(Uri url)
---> (Inner Exception #0) System.Threading.Tasks.TaskCanceledException: A task was canceled.<---
```

> It's more verbose this time.

But it looks like it continued and finished the update this time?

**Comment by Falki-git** [comment]
Yes, it completed the update

**Comment by HebaruSan** [comment]
Alright, based on that and the fact that the previous implementation seemingly never had this problem, I have to consider the possibility that `HttpClient` just doesn't work very well (despite the fact that it has a [compiler warning recommending its use](https://learn.microsoft.com/en-us/dotnet/fundamentals/syslib-diagnostics/syslib0014)). To provide a point of comparison, here's a fresh build with the old pre-`HttpClient` implementation of `CurrentETag`:

- [ckan.zip](https://github.com/KSP-CKAN/CKAN/files/13697673/ckan.zip)

If that build works for you, @Falki-git and @dandoesgithub, without a 100-second delay, then we'll probably need to roll back the switch to `HttpClient`. It would be nicer to keep up with the latest and greatest library code, but if `HttpClient` is just a hopelessly buggy mess, then we can't reasonably use it.

**Comment by dandoesgithub** [comment]
> @dandoesgithub and @Falki-git, here's a new test build for you to try:
> 
>     * (file removed)
> 
> 
> This one does new things: first it prints the value of `HttpClient.Timeout`, and then it catches, logs, and ignores any exceptions it throws while trying to get the etags. This _**might**_ help with the problem you're having, but whether it does or not, it's important that we get the log file so we can determine the right long term fix for the real builds.
> 
> > Regarding the video, I should have been explicit also, I cut it down to 15 seconds so you don't need to look at a static screen, but from clicking Refresh to the timeout it takes a minute or so (likely 100 secs as you said).
> 
> Thanks, that makes a lot more sense! That makes me a lot more confident concluding that this is in fact `HttpClient` timing out. I built and typed the above before you posted this, but I think it's still worth capturing those logs.

sorry for the late reply, just noticed you sent a new ckan.exe. i'll send the logs of the one you sent before, then the logs of  the one rn

**Comment by dandoesgithub** [comment]
well, this build definitely works when refreshing my instance. 
[CKAN.log](https://github.com/KSP-CKAN/CKAN/files/13697688/CKAN.log)

**Comment by HebaruSan** [comment]
Thanks, that's one vote solidly in favor of reverting to `WebRequest`, then. If Falki sees the same thing, `HttpClient` will be escorted from the premises.

**Comment by Falki-git** [comment]
Yes, that's a lot better. I cannot reproduce the timeout anymore. Tried multiple instances, refreshed many times, it works 100% now.
[CKAN_logs4.zip](https://github.com/KSP-CKAN/CKAN/files/13697706/CKAN_logs4.zip)

**Comment by Dominiquini** [comment]
I fix this bug removing the [Kopernicus_BE] repository!

**Comment by HebaruSan** [comment]
Hi @Dominiquini, this is an inconsistent, intermittent bug, so it is easy to mistake a coincidence for a cause. The other people having this issue weren't even using that repo. FYI that I'm going to hide your comment to avoid confusing other users who may land here. Cheers!

**Comment by Dominiquini** [comment]
> Hi @Dominiquini, this is an inconsistent, intermittent bug, so it is easy to mistake a coincidence for a cause. The other people having this issue weren't even using that repo. FYI that I'm going to hide your comment to avoid confusing other users who may land here. Cheers!

In my case, I was able to reproduce it very consistently (adding and removing this repository)!

Thanks

**Comment by HebaruSan** [comment]
@Dominiquini, no, you saw it happen and you assumed it had to do with the repository. We investigated this, and the code that is throwing the exception has nothing to do with that. Please have some consideration for other users who may see your comments and be misled.

**Comment by Dominiquini** [comment]
> @Dominiquini, no, you saw it happen and you assumed it had to do with the repository. We investigated this, and the code that is throwing the exception has nothing to do with that. Please have some consideration for other users who may see your comments and be misled.

I'm only reporting my experience! I even hid the comment! How can you say that I have no consideration for other users?
You marked that my bug was a duplicate of this one! In my case, I tested it about 4 times with and without the repository and the result was consistent!

**Comment by HebaruSan** [comment]
Hi @Dominiquini, this issue is not about a back and forth between me and you, it is about `HttpClient` throwing a `TaskCanceledException` after a bogus timeout that gets wrapped in a `fromAggregateException` and reported as "One or more errors occurred". I have locked the conversation and deleted the off-topic comments. Thanks for filing the duplicate bug report.
